### ISSUE TITLE
Failing evaluation of constant not reported in cargo check

### ISSUE THREAD

**Issue by hniksic** [rust-lang/rust#116936]
This function rejects zero-sized types at compile-time. (That's a minimal example, real code would check other properties of the type not expressed via a trait, e.g. its alignment, etc., for unsafe/FFI/serialization purposes.)

```rust
fn reject_zst<T>(_value: T) {
    struct Check<T>(T);
    impl<T> Check<T> {
        const NON_ZERO_SIZE: () = assert!(std::mem::size_of::<T>() != 0);
    }
    let _ = Check::<T>::NON_ZERO_SIZE;
}

fn main() {
    reject_zst(3.14f64); // compiles
    reject_zst(()); // doesn't compile
}
```

With `cargo build` the code is rejected, as expected. However with `cargo check` it passes compilation.

I expected the compilation to fail with both `cargo check` and `cargo build`.

Tested in Rust 1.71.0.

**Comment by bjorn3** [comment]
`cargo check` doesn't report post-monomorphization errors like this. This is tracked at several places like https://github.com/rust-lang/rust/issues/99682. I believe there was a discussion about if we should guarantee that all post-monomorphization will result in an error for `cargo check` or not, but I can't find it anymore.

**Comment by RalfJung** [comment]
Closing as duplicate of https://github.com/rust-lang/rust/issues/99682 -- and also as not-a-bug; this is currently intended behavior since checking these constants requires doing a monomorphization pass which would massively slow down the build.
